packages/xod-client/src/hinting/typeDeduction.js

```diff
@@ -1,21 +1,74 @@
 import * as R from 'ramda';
 import * as XP from 'xod-project';
-import { isAmong, foldMaybe, maybeProp, maybePath } from 'xod-func-tools';
+import {
+  isAmong,
+  foldMaybe,
+  foldMaybeWith,
+  maybeProp,
+  maybePath,
+  explodeMaybe,
+} from 'xod-func-tools';
 
 import * as PAT from '../project/actionTypes';
 
-import { getActingPatchPath } from './utils';
+import { getActingPatchPath } from './errorCollectors';
 
 // =============================================================================
 //
 // Type deduction
 //
 // =============================================================================
 
-// Run type deduction only for some actions
-// :: Project -> Action -> Boolean
+/**
+ * Returns True if pin types should be deduced and False if not.
+ *
+ * Run type deduction for all actions, except actions that are listed here
+ * or that does not pass any other checks.
+ *
+ * shallDeduceTypes :: Project -> Action -> Boolean
+ */
 export const shallDeduceTypes = R.curry((project, action) =>
   R.anyPass([
+    // Blacklist
+    // Do not deduce for these action types:
+    R.propSatisfies(
+      R.complement(
+        isAmong([
+          PAT.PROJECT_OPEN_REQUESTED,
+          PAT.PROJECT_PUBLISH_REQUESTED,
+          PAT.PROJECT_PUBLISH_CANCELLED,
+          PAT.PROJECT_PUBLISH_START,
+          PAT.PROJECT_PUBLISH_SUCCESS,
+          PAT.PROJECT_PUBLISH_FAIL,
+          PAT.PROJECT_OPEN_WORKSPACE,
```

Comment:
Do the actions from the head ever affect the `project` branch in the Redux state? Namely:

```
          PAT.PROJECT_OPEN_REQUESTED,
          PAT.PROJECT_PUBLISH_REQUESTED,
          PAT.PROJECT_PUBLISH_CANCELLED,
          PAT.PROJECT_PUBLISH_START,
          PAT.PROJECT_PUBLISH_SUCCESS,
          PAT.PROJECT_PUBLISH_FAIL,
          PAT.PROJECT_OPEN_WORKSPACE,
```

I mean, if they do not change `project`, no reason to list them here because the check will never be run.